### PR TITLE
Upgrade Khoj Obsidian: Chat from Side Pane, Stream Intermediate Steps, Copy Message to Clipboard

### LINKED ISSUE #789: [FIX] Fix Styling, Selectable Text and Exiting Chat View in Obsidian
## Describe the bug
- Formatting of Bullet List, Code block and Selectable Text in Obsidian
- Chat view in Obsidian

## To Reproduce
- Make Khoj generate a bulleted list in response
- Click Chat view with Conversation Session in Chat view open

## Screenshots
N/A

## Platform
- Server:
    - [X] Cloud-Hosted (https://app.khoj.dev)
    - [X] Self-Hosted Docker
    - [X] Self-Hosted Python package
    - [X] Self-Hosted source code
- Client:
    - [X] Obsidian
    - [ ] Emacs
    - [ ] Desktop app
    - [ ] Web browser
    - [ ] WhatsApp
- OS:
    - [X] Windows
    - [X] macOS
    - [X] Linux
    - [ ] Android
    - [ ] iOS

### If self-hosted
- Server Version [e.g. 1.0.1]:

### PR DESCRIPTION
**Pull request by debanjum**
### Details
- **Chat with Khoj from right pane on Obsidian**
  - Modal was too ephemeral, couldn't have it open for reference, quick jump to Khoj chat
- **Stream intermediate steps taken by Khoj** for generating response to the chat pane
  Gives more transparency into Khoj 'thinking' process, e.g internet, notes searches performed, documents read etc. 
  The feedback allows us to tune our messages to elicit better responses by Khoj
- Add ability to **copy message to clipboard, paste chat messages directly into current file**
- Jump to **Search**, **Find Similar** functions from navigation bar on the Khoj Obsidian side pane
- Improve spacing, use consistent colors in chat message references and buttons
### Screenshots
| Type | View |
|--------|--------|
| Khoj in Obsidian Side Pane | Khoj Chat in Obsidian Side Pane |
| Chat Sessions in Obsidian Side Pane | Obsidian Conversation Sessions | 

Resolves #789, #754

### CODE REVIEW

**Review by sabaimran — COMMENTED** [review]
These improvements are super exciting for the Obsidian interface ✨

**Comment by TaiTaiTai777** [comment]
### Enhancement Proposal for Khoj in Obsidian

To enhance user experience and functionality, the following improvements are recommended for the Khoj integration with Obsidian:

#### 1. Persistent Chat Modal

**Current Issue**:
- The chat modal is too ephemeral and does not stay open for quick reference.

**Proposed Solution**:
- Implement a persistent chat modal within the right pane of Obsidian that remains open until explicitly closed by the user. This ensures users can maintain a continuous reference without losing context.

---

#### 2. Intermediate Steps Transparency

**Current Issue**:
- Lack of transparency in the intermediate steps taken by Khoj to generate a response.

**Proposed Solution**:
- Enable a feature that streams the intermediate steps Khoj takes for generating each response. This includes:
  - Internet searches performed.
  - Notes and documents read or referenced.
  - Search indexes queried.

**Benefits**:
- Provides users insight into Khoj’s reasoning process.
- Helps users tailor their queries for more accurate and relevant responses.

---

#### 3. Copy and Paste Functionality

**Current Issue**:
- Lack of ability to copy messages to the clipboard or paste chat messages directly into the current file.

**Proposed Solution**:
- Add a button to each chat message allowing users to copy the message to the clipboard.
- Enable a "Paste to Current File" button which directly inserts the selected chat message into the currently open note in Obsidian.

**Benefits**:
- Streamlines the workflow by reducing the need for manual copying and pasting.
- Enhances productivity by allowing rapid integration of chat responses into notes.

---

#### 4. Enhanced Navigation Functions

**Current Issue**:
- Navigation functions like "Jump to Search" and "Find Similar" are not easily accessible.

**Proposed Solution**:
- Integrate "Jump to Search" and "Find Similar" functions into the navigation bar on the Khoj side pane in Obsidian.

**Benefits**:
- Provides quick access to search-related functionalities.
- Improves user experience by reducing the steps required to perform searches and access related notes.

---

#### 5. Improved UI and Consistency

**Current Issue**:
- Inconsistent spacing and color use in chat message references and buttons.

**Proposed Solution**:
- Standardize spacing and color palettes across all chat messages and buttons to maintain consistency.
- Ensure a minimalistic and clean design that aligns with Obsidian’s overall aesthetics.

**Benefits**:
- Enhances visual clarity and appeal.
- Provides a consistent user experience which can reduce cognitive load and improve usability.

---

### Example Implementations

#### Persistent Chat Modal

**UI Concept**:
- A detachable chat window that sticks to the right pane, with a clear "Close" button.

```html
<div class="persistent-modal">
  <div class="modal-header">
    <h3>Khoj Chat</h3>
    <button class="close-btn">&times;</button>
  </div>
  <div class="modal-content">
    <!-- Chat content here -->
  </div>
</div>
```

**CSS**:
```css
.persistent-modal {
  position: fixed;
  top: 0;
  right: 0;
  width: 300px;
  height: 100%;
  background-color: #fff;
  box-shadow: -2px 0 5px rgba(0,0,0,0.1);
}
.close-btn {
  background: transparent;
  border: none;
  font-size: 20px;
  cursor: pointer;
}
```

---

#### Intermediate Steps Transparency

**UI Concept**:
- A collapsible section below chat responses showing intermediate steps.

```html
<div class="chat-response">
  <p>{Khoj Response}</p>
  <div class="intermediate-steps">
    <button class="toggle-steps">Show Steps</button>
    <div class="steps-content" style="display:none;">
      <!-- Steps shown here -->
    </div>
  </div>
</div>
```

**JavaScript**:
```javascript
document.querySelector('.toggle-steps').addEventListener('click', function() {
  const content = this.nextElementSibling;
  content.style.display = content.style.display === 'none' ? 'block' : 'none';
  this.textContent = content.style.display === 'none' ? 'Show Steps' : 'Hide Steps';
});
```

---

#### Copy and Paste Functionality

**UI Concept**:
- Adding copy and paste buttons alongside each message.

```html
<div class="chat-message">
  <p>{Chat Message}</p>
  <button class="copy-btn">Copy</button>
  <button class="paste-btn">Paste to File</button>
</div>
```

**JavaScript**:
```javascript
document.querySelector('.copy-btn').addEventListener('click', function() {
  const message = this.previousElementSibling.innerText;
  navigator.clipboard.writeText(message);
});

document.querySelector('.paste-btn').addEventListener('click', function() {
  const message = this.previousElementSibling.innerText;
  // Insert message into the current file in Obsidian logic here.
});
```

---

By implementing these enhancements, Khoj will provide a more robust, user-friendly, and efficient experience within Obsidian, fostering better interaction and productivity.

**Review by lentinidante — APPROVED** [review]
lgtm. changes ensure the chat input field is focused and scrolls to the latest message efficiently using requestAnimationFrame.

File: src/interface/obsidian/src/chat_view.ts [hunk 1]
```diff
@@ -0,0 +1,1306 @@
...
+        }
+
+        // Get chat response from Khoj backend
+        let encodedQuery = encodeURIComponent(query);
+        let chatUrl = `${this.setting.khojUrl}/api/chat?q=${encodedQuery}&n=${this.setting.resultsCount}&client=obsidian&stream=true&region=${this.location.region}&city=${this.location.city}&country=${this.location.countryName}&timezone=${this.location.timezone}`;
+        let responseElement = this.createKhojResponseDiv();
+
+        // Temporary status message to indicate that Khoj is thinking
+        this.result = "";
+        let loadingEllipsis = this.createLoadingEllipse();
```
Comment (sabaimran) [inline src/interface/obsidian/src/chat_view.ts:756]: the loading ellipse doesn't seem to be showing up for me, but not sure why.

File: src/interface/obsidian/src/chat_view.ts [hunk 2]
```diff
@@ -0,0 +1,1306 @@
...
+
+    addMessageToChatBody(rawResponse: string, newResponseElement: HTMLElement | null, references: any) {
+        if (!newResponseElement) return;
+        newResponseElement.innerHTML = "";
+        newResponseElement.appendChild(this.formatHTMLMessage(rawResponse));
+
+        this.finalizeChatBodyResponse(references, newResponseElement);
+    }
+
+    finalizeChatBodyResponse(references: object, newResponseElement: HTMLElement | null) {
```
Comment (sabaimran) [inline src/interface/obsidian/src/chat_view.ts:1094]: Make sure to add the copy/insert buttons here too! Super neat to use those.

**Review by sabaimran — COMMENTED** [review]
Just a few more suggestions, but the experience including sessions is so much better.

File: src/interface/obsidian/src/chat_view.ts [hunk 3]
```diff
@@ -0,0 +1,1306 @@
...
+
+        // Add chat input field
+        let inputRow = contentEl.createDiv("khoj-input-row");
+        let chatSessions = inputRow.createEl("button", {
+            text: "Chat Sessions",
+            attr: {
+                class: "khoj-input-row-button clickable-icon",
+            },
+        })
+        chatSessions.addEventListener('click', async (_) => { await this.showChatSessions(chatBodyEl) });
```
[the remark this answers is not in the document]
Comment (debanjum) [reply src/interface/obsidian/src/chat_view.ts]: Done

File: src/interface/obsidian/src/chat_view.ts (hunk 1 above)
Comment (debanjum) [reply src/interface/obsidian/src/chat_view.ts:759]: Yeah, the loading ellipsis needed to be of a contrasting color. Fixed that

File: src/interface/obsidian/src/chat_view.ts [hunk 4]
```diff
@@ -0,0 +1,1306 @@
...
+            });
+
+    }
+
+    getViewType(): string {
+        return KhojView.CHAT;
+    }
+
+    getDisplayText(): string {
+        return "Khoj Chat";
```
[the remark this answers is not in the document]
Comment (debanjum) [reply src/interface/obsidian/src/chat_view.ts:56]: `getDisplayText` is expected to return a string that's shown on hover (I think). Putting in the host url here maybe strange?

File: src/interface/obsidian/src/chat_view.ts [hunk 5]
```diff
@@ -0,0 +1,1306 @@
...
+                                this.handleStreamResponse(this.websocketState.newResponseTextEl, chunk.message, this.websocketState.loadingEllipsis, true);
+                            } else {
+                                this.websocketState.rawResponse = chunk.response;
+                            }
+                        } catch (error) {
+                            // If the chunk is not a JSON object, just display it as is
+                            this.websocketState.rawResponse += chunk;
+                        } finally {
+                            if (chunk.type != "status" && chunk.type != "rate_limit") {
+                                this.addMessageToChatBody(this.websocketState.rawResponse, this.websocketState.newResponseTextEl, this.websocketState.references);
```
[the remark this answers is not in the document]
Comment (debanjum) [reply src/interface/obsidian/src/chat_view.ts]: That is super strange. I don't have see it on my end but given it's happening independent of this change. Let's keep the debugging and fixing of this issue separate

**Pull request merged**